### PR TITLE
Post-PalEON updates

### PR DESCRIPTION
**Pull request by istfer**
Changes to PDA code to address the identified issues during PalEON and to bring other algorithms (other than emulator) up-to-date

- Fixed the time mismatch in load_x.netcdf due to time zone differences and made corresponding changes in the PDA code
- Instead of saving only the diagonals, now I'm saving the whole jcov matrix for the extension runs (to run the emulator chains longer)
- The rest is mostly about making the other functions work with the latest developments (running them for more than one PFTs and for more than one chain

I believe @crollinson is up for review

### CODE REVIEW

File: modules/uncertainty/R/run.ensemble.analysis.R [hunk 1]
```diff
@@ -131,8 +131,9 @@ run.ensemble.analysis <- function(settings,plot.timeseries=NA, ensemble.id=NULL,
 ##'
 ##' @author Michael Dietze, Ryan Kelly
 ##'
-read.ensemble.ts <- function(model, ensemble.id=NULL, variable=NULL, start.year=NULL, end.year=NULL) {
+read.ensemble.ts <- function(settings, ensemble.id=NULL, variable=NULL, start.year=NULL, end.year=NULL) {
```
Comment (istfer) [inline modules/uncertainty/R/run.ensemble.analysis.R]: @mdietze the only function that calls read.ensemble.ts besides the workflow is sda.particle, I haven't modified it

File: modules/assim.batch/R/pda.emulator.R [hunk 2]
```diff
@@ -218,20 +221,23 @@ pda.emulator <- function(settings, params.id=NULL, param.names=NULL, prior.id=NU
     }else{
       load(settings$assim.batch$emulator.path) # load previously built emulator to run a longer mcmc
```
Comment (istfer) [inline modules/assim.batch/R/pda.emulator.R]: I might put emulator and llik under resume path too, for now left them separate

File: modules/assim.batch/R/pda.emulator.R [hunk 3]
```diff
@@ -345,23 +353,24 @@ pda.emulator <- function(settings, params.id=NULL, param.names=NULL, prior.id=NU
...
-    # merge with previous run's mcmc samples
-    mcmc.list <- mapply(rbind, mcmc.list, mcmc.list.tmp, SIMPLIFY=FALSE)
-    settings$assim.batch$iter <- nrow(mcmc.list[[1]])
-    
-  }else{
-    
-    mcmc.list <- mcmc.list.tmp
-
-  }
+  # if(length(extension.check)==1 & !run.block){
```
Comment (istfer) [inline modules/assim.batch/R/pda.emulator.R]: commented out some lines here and there while updating, maybe I better delete these

File: modules/assim.batch/vignettes/AssimBatchVignette.Rmd [hunk 4]
```diff
@@ -194,23 +211,26 @@ If you are using methods other than *bruteforce* and *bruteforce.bs*, some addit
 * DEzs : Differential Evolution with a snooker updater
 * DREAM : Differential Evolution Adaptive Metropolis 
 * DREAMzs : Differential Evolution Adaptive Metropolis with a snooker updater
+* Twalk : "traverse" or "thoughtful" walk, a general purpose sampling algorithm
 * SMC : Sequential Monte Carlo
 
 The name of the chosen sampler would be passed under `<sampler>` tag within the `<bt.settings>` block :
 
+It is not possible to use some of the samplers in this package for univariate cases. The ones that you can use for univariate cases are: "Metropolis", "DE", "DEzs" and "Twalk".
```
Comment (istfer) [inline modules/assim.batch/vignettes/AssimBatchVignette.Rmd]: I'll work on BayesianTools code and usage some more later. I'd like to implement the parallelization

File: modules/assim.batch/R/pda.emulator.R [hunk 5]
```diff
@@ -184,6 +184,9 @@ pda.emulator <- function(settings, params.id=NULL, param.names=NULL, prior.id=NU
   ## if it is not specified, default to GPfit
```
Comment (mccabete) [inline modules/assim.batch/R/pda.emulator.R]: What does the "it" in the line 184 comment refer too?

File: modules/assim.batch/R/pda.emulator.R [hunk 6]
```diff
@@ -218,20 +221,23 @@ pda.emulator <- function(settings, params.id=NULL, param.names=NULL, prior.id=NU
...
-        
-        prior.all <- do.call("rbind", prior.list)
-        prior.ind.all <- which(unlist(pname) %in% unlist(settings$assim.batch$param.names))
-        prior.fn.all <- pda.define.prior.fn(prior.all)
-        
-        init.list[[c]] <-  as.list(sapply(seq_along(prior.ind.all), 
-                                          function(x) eval(prior.fn.all$pprior[[prior.ind.all[x]]], list(q=init.x[x]))))
+        init.list[[c]] <- resume.list[[c]]$prev.samp[nrow(resume.list[[c]]$prev.samp),]
+        jmp.list[[c]] <- resume.list[[c]]$jump
+        # init.x <- mcmc.list[[c]][nrow(mcmc.list[[c]]),]
```
Comment (mccabete) [inline modules/assim.batch/R/pda.emulator.R]: Is this a comment left over from testing? It looks like it was recently commented out and recently edited?

Comment (istfer) [inline modules/assim.batch/R/pda.emulator.R]: yes, it was left from the previous implementation, now I erased the unused commented-out code bits

File: modules/assim.batch/R/pda.mcmc.R [hunk 7]
```diff
@@ -72,153 +74,170 @@ pda.mcmc <- function(settings, params.id=NULL, param.names=NULL, prior.id=NULL,
   settings$assim.batch$ensemble.id <- pda.create.ensemble(settings, con, workflow.id)
 
   ## Set prior distribution functions (d___, q___, r___, and multivariate versions)
-  prior.fn <- pda.define.prior.fn(prior)
+  #prior.fn <- lapply(prior.list, pda.define.prior.fn)
```
Comment (mccabete) [inline modules/assim.batch/R/pda.mcmc.R]: Again, this looks like it's leftover from testing?

File: modules/assim.batch/R/pda.mcmc.R [hunk 8]
```diff
@@ -72,153 +74,170 @@ pda.mcmc <- function(settings, params.id=NULL, param.names=NULL, prior.id=NULL,
...
+  pname.all <- unlist(pname)
 
   ## ----------------------------------- MCMC Setup ----------------------------------- ##
   ## Initialize empty params matrix (concatenated to params from a previous PDA, if provided)
-  params <- pda.init.params(settings, con, pname, n.param.all)
-    start  <- params$start
-    finish <- params$finish
-    params <- params$params
 
   ## File for temp storage of params (in case of crash)
```
Comment (mccabete) [inline modules/assim.batch/R/pda.mcmc.R]: What lines are the comments from 91-93 referring to?

File: modules/assim.batch/R/pda.mcmc.bs.R [hunk 9]
```diff
@@ -71,156 +72,160 @@ pda.mcmc.bs <- function(settings, params.id=NULL, param.names=NULL, prior.id=NUL
...
+  mcmc.list <- jcov.list <- list()
+  
+  for(chain in 1:settings$assim.batch$chain){
+    ## Initialize empty params matrix (concatenated to params from a previous PDA, if provided)
+    params.list <- pda.init.params(settings, chain, pname.all, sum(n.param.all))
+    start  <- params.list$start
+    finish <- params.list$finish
+    params <- params.list$params
+    
+    ## File for temp storage of params (in case of crash)
```
Comment (mccabete) [inline modules/assim.batch/R/pda.mcmc.bs.R]: Already commented on this block of comments earlier, just flagging it again.

File: modules/assim.batch/R/pda.mcmc.R (hunk 8 above)
Comment (istfer) [inline modules/assim.batch/R/pda.mcmc.R]: it was for a previous way of saving the chain, now deleted

File: modules/benchmark/R/load.netcdf.R [hunk 10]
```diff
@@ -50,19 +50,19 @@ load.x_netcdf <- function(data.path, format, site, vars=NULL){
     # throw error if can't parse time format  
     if(is.na(date.origin)) logger.error("All time formats failed to parse. No formats found.")
     
-    time.stamp.match <- gsub("UTC", "", date.origin)
+    time.stamp.match <- gsub("UTC", "", date.origin) 
     t.units <- gsub(paste0(" since ", time.stamp.match, ".*"), "", ncatt_get(nc[[i]], dims[time.var])$units)
```
Comment (istfer) [inline modules/benchmark/R/load.netcdf.R]: according to stackoverflow there is a lot of story behind it :) https://tools.ietf.org/html/rfc3092
here writing these in two lines just helped me while testing and debugging, could be reduced to one line indeed

**Pull request merged**